include/ghex/unstructured/communication_object_ipr.hpp

```diff
@@ -0,0 +1,561 @@
+                    using buffer_info_type = buffer_info<pattern_type,D,F>;
+
+                private:
+
+                    // member types
+
+                    using address_type = typename communicator_type::address_type;
+                    using index_container_type = typename pattern_type::index_container_type;
+                    using pack_function_type = std::function<void(void*, const index_container_type&, void*)>; // last argument intended as CUDA stream
+
+                    /** @brief pair of domain ids + tag id with ordering */
+                    struct domain_id_pair_and_tag {
+
+                            domain_id_type first_id;
+                            domain_id_type second_id;
+                            int tag;
+
+                            bool operator<(const domain_id_pair_and_tag& other) const noexcept {
+                                return (first_id < other.first_id ? true :
+                                    (first_id > other.first_id ? false :
+                                        (second_id < other.second_id ? true :
+                                            (second_id > other.second_id ? false :
+                                                (tag < other.tag)))));
+                            }
+
+                    };
+
+                    /** @brief Holds a pointer to a set of iteration spaces and a callback function pointer
+                     * which is used to store a field's pack or unpack member function.
+                     * This class also stores the offset in the serialized buffer in bytes.
+                     * The type-erased field_ptr member is only used for the gpu-vector-interface.
+                     * @tparam Function either pack or unpack function pointer type */
+                    template<typename Function>
+                    struct field_info {
+                            using index_container_type = typename pattern_type::map_type::mapped_type;
+                            Function call_back;
+                            const index_container_type* index_container;
+                            std::size_t offset;
+                            void* field_ptr;
+                    };
+
+                    /** @brief Holds serial buffer memory and meta information associated with it.
+                     * @tparam Vector contiguous buffer memory type
+                     * @tparam Function pack function pointer type (not used for unpacking)*/
+                    template<class Vector, class Function>
+                    struct buffer {
+                            using field_info_type = field_info<Function>;
+                            address_type address;
+                            int tag;
+                            Vector buffer;
+                            std::size_t size;
+                            std::vector<field_info_type> field_infos;
+                            cuda::stream m_cuda_stream;
```

Comment:
Is it ok to stick with this naming convention (same used in `communication_object_2`)?

Reply:
I did not notice the names here. We use `m_` for members that are not assumed to be accessed publicly.
Anyway, since these members are not accessed in this PR, I would leave the code as is, since otherwise you have to make more diffs.

Reply:
Ok, sure